ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/StoreBackedCommandExecutor.scala

```diff
@@ -168,40 +177,48 @@ private[apiserver] final class StoreBackedCommandExecutor(
-            .future(
-              metrics.daml.execution.lookupActiveContract,
-              contractStore.lookupActiveContract(readers, acoid),
-            )
-            .flatMap { instance =>
-              lookupActiveContractTime.addAndGet(System.nanoTime() - start)
-              lookupActiveContractCount.incrementAndGet()
-              resolveStep(
-                Timed.trackedValue(
-                  metrics.daml.execution.engineRunning,
-                  trackSyncExecution(interpretationTimeNanos)(resume(instance)),
+          disclosedContractsByContractId
+            .get(acoid)
+            .fold {
+              val start = System.nanoTime
+              Timed
+                .future(
+                  metrics.daml.execution.lookupActiveContract,
+                  contractStore.lookupActiveContract(readers, acoid),
                 )
-              )
-            }
+                .flatMap { instance =>
+                  lookupActiveContractTime.addAndGet(System.nanoTime() - start)
+                  lookupActiveContractCount.incrementAndGet()
+                  resolveStep(
+                    Timed.trackedValue(
+                      metrics.daml.execution.engineRunning,
+                      trackSyncExecution(interpretationTimeNanos)(resume(instance)),
+                    )
+                  )
+                }
+            }(contract => resolveStep(resume(Some(contract))))
 
         case ResultNeedKey(key, resume) =>
-          val start = System.nanoTime
-          Timed
-            .future(
-              metrics.daml.execution.lookupContractKey,
-              contractStore.lookupContractKey(readers, key.globalKey),
-            )
-            .flatMap { contractId =>
-              lookupContractKeyTime.addAndGet(System.nanoTime() - start)
-              lookupContractKeyCount.incrementAndGet()
-              resolveStep(
-                Timed.trackedValue(
-                  metrics.daml.execution.engineRunning,
-                  trackSyncExecution(interpretationTimeNanos)(resume(contractId)),
+          disclosedContractsByKeyHash
+            .get(key.globalKey.hash)
+            .fold {
+              val start = System.nanoTime
+              Timed
+                .future(
+                  metrics.daml.execution.lookupContractKey,
+                  contractStore.lookupContractKey(readers, key.globalKey),
                 )
-              )
-            }
+                .flatMap { contractId =>
+                  lookupContractKeyTime.addAndGet(System.nanoTime() - start)
+                  lookupContractKeyCount.incrementAndGet()
+                  resolveStep(
+                    Timed.trackedValue(
+                      metrics.daml.execution.engineRunning,
+                      trackSyncExecution(interpretationTimeNanos)(resume(contractId)),
+                    )
+                  )
+                }
+            }(contractId => resolveStep(resume(Some(contractId))))
```

Comment:
note that for this case we want to ask the Engine to validate that the key was actually correct to fail early if users fake the key hash. Nothing to do yet until @robin-da implements the API (probably a second parameter to `resume`). maybe consider adding a TODO in the meanwhile.

Reply:
Added TODOs related to https://github.com/digital-asset/daml/issues/13848